Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -621,9 +621,66 @@ However, this is not a concern on iOS. As discussed in the section [Testing on t
 
 #### Overview
 
-Obfuscation is the process of transforming code and data to make it more difficult to comprehend. It is an integral part of every software protection scheme. What's important to understand is that obfuscation isn't something that can be simply turned on or off. Programs can be made incomprehensible, in whole or in part, in many ways and to different degrees.
+Obfuscation is a process of transforming the original algorithms, data structures into a form that is difficult to disassemble and understand but the program after the modification preserves the original functionality. Both code and data can be obfuscated, so that an attacker can mostly rely on dynamic analysis.
 
-#### SwiftShield
+> Note: All presented techniques below may not stop reverse engineers, but combining all of those techniques will make their job significantly harder. The aim of those techniques is to discourage reverse engineers from further analysis.
+
+#### Obfuscation Techniques
+
+The following techniques can be used to obfuscate an application:
+- Name obfuscation of classes, methods and fields
+- Instructions substitution
+- Control flow flattening
+- Dead code injection
+- String encryption
+
+#### Name obfuscation of classes, methods and fields
+
+The standard compiler generates binary’s symbols based on class and function names from the source code. Therefore, symbol names remain meaningful and can be easily read by disassembling the application. For instance, a function which detects a jailbreak can be located by searching for relevant keywords inside the binary. The listing below shows the disassembled function *JailbreakDetectionViewController.jailbreakTest4Tapped(..)* from DVIA-v2 app.
+
+```assembly
+__T07DVIA_v232JailbreakDetectionViewControllerC20jailbreakTest4TappedyypF:
+stp        x22, x21, [sp, #-0x30]!
+mov        rbp, rsp
+```
+
+After the obfuscation we can observe that the symbol’s name is no longer meaningful as shown on the listing below.
+
+```assembly
+__T07DVIA_v232zNNtWKQptikYUBNBgfFVMjSkvRdhhnbyyFySbyypF:
+stp        x22, x21, [sp, #-0x30]!
+mov        rbp, rsp
+```
+
+Nevertheless, the binary’s code is not modified, so an attacker can still retrieve the logic of the algorithm by reading the disassembled version of the function and figure out its purpose.
+
+#### Instructions substitution
+
+This technique replaces standard binary operators like addition, subtraction with more complex representations. For example an addition `x = a + b` can be represented as `x = -(-a) - (-b)`. However, using the same replacement representation could be easily reversed, so it is recommended to add multiple substitution techniques for a single case and introduce a random factor. Nevertheless, this technique is vulnerable to code re-optimaliztion.
```

Comment:
```suggestion
This technique replaces standard binary operators like addition or subtraction with more complex representations. For example an addition `x = a + b` can be represented as `x = -(-a) - (-b)`. However, using the same replacement representation could be easily reversed, so it is recommended to add multiple substitution techniques for a single case and introduce a random factor. Nevertheless, this technique is vulnerable to de-obfuscation, but it can still be time consuming depending on the complexity and depth of the substitutions.
```